File: simulacion_iot.js
// ================= CONFIG =================
const BROKER = "wss://test.mosquitto.org:8081/mqtt";
const TOPIC = "sensor/gasolina";
const UMBRAL_GAS = 300;

let simulando = true;

// ================= MQTT =================
const client = mqtt.connect(BROKER);

// 🔊 desbloquear audio en navegador
document.addEventListener("click", () => {
    const ctx = new (window.AudioContext || window.webkitAudioContext)();
    ctx.resume();
});

// ================= CONEXIÓN =================
client.on("connect", () => {
    console.log("✅ Conectado al broker MQTT");
    client.subscribe(TOPIC);
});

// ================= RECEPCIÓN =================
client.on("message", (topic, message) => {
    let valor = parseFloat(message.toString());

    if (isNaN(valor)) return;

    console.log("📥 Dato recibido:", valor);

    if (valor > UMBRAL_GAS) {
        console.log("🚨 ALERTA: Posible fuga de gasolina 🚨");
        alert("🚨 FUGA DETECTADA: " + valor);
        beep();
    }
});

// ================= SIMULADOR (CORREGIDO) =================
setInterval(() => {

    if (!simulando) return;

    let valor = Math.floor(Math.random() * 400) + 100;

    console.log("📤 Enviando dato:", valor);

    client.publish(TOPIC, valor.toString());

    // ⚠️ esto evita errores si existe función en HTML
    if (typeof procesar === "function") {
        procesar(valor);
    }

}, 2000);

console.log("🚀 Simulación iniciada...");

// ================= BEEP =================
function beep() {
    try {
        const ctx = new (window.AudioContext || window.webkitAudioContext)();

        const osc = ctx.createOscillator();
        osc.type = "square";
        osc.frequency.value = 1200;

        osc.connect(ctx.destination);
        osc.start();

        setTimeout(() => {
            osc.stop();
        }, 200);

    } catch (e) {
        console.log("Audio error:", e);
    }
}

// ================= QR CLOSE =================
function cerrarQR() {
    const modal = document.getElementById("qrModal");
    if (modal) {
        modal.style.display = "none";
    }
}

File: simulacion_iot.py
import random
import time
import threading
import paho.mqtt.client as mqtt

# Configuración
BROKER = "broker.hivemq.com"
TOPIC = "sensor/gasolina"
UMBRAL_GAS = 300

client = mqtt.Client()

def on_connect(client, userdata, flags, rc):
    print("✅ Conectado al broker MQTT")
    client.subscribe(TOPIC)

def on_message(client, userdata, msg):
    valor = float(msg.payload.decode())
    print(f"📥 Dato recibido: {valor}")

    if valor > UMBRAL_GAS:
        print("🚨 ALERTA: Posible fuga de gasolina 🚨")

client.on_connect = on_connect
client.on_message = on_message

client.connect(BROKER, 1883, 60)

def simular_sensor():
    print("📡 Sensor iniciado...")
    while True:
        valor = random.randint(100, 500)
        print(f"📤 Enviando dato: {valor}")
        client.publish(TOPIC, valor)
        time.sleep(5)

# Hilo
thread = threading.Thread(target=simular_sensor)
thread.daemon = True
thread.start()

print("🚀 Simulación iniciada...")

client.loop_forever()